src/libraries/System.IO.Compression/src/System/IO/Compression/ZipCustomStreams.cs

```diff
@@ -366,7 +373,28 @@ async ValueTask<int> Core(Memory<byte> buffer, CancellationToken cancellationTok
         public override long Seek(long offset, SeekOrigin origin)
         {
             ThrowIfDisposed();
-            throw new NotSupportedException(SR.SeekingNotSupported);
+
+            if (!CanSeek)
+            {
+                throw new NotSupportedException(SR.SeekingNotSupported);
+            }
+
+            long newPositionInSuperStream = origin switch
+            {
+                SeekOrigin.Begin => _startInSuperStream + offset,
+                SeekOrigin.Current => _positionInSuperStream + offset,
+                SeekOrigin.End => _endInSuperStream + offset,
+                _ => throw new ArgumentOutOfRangeException(nameof(origin)),
+            };
+
+            if (newPositionInSuperStream < _startInSuperStream)
+            {
+                throw new IOException(SR.IO_SeekBeforeBegin);
+            }
+
+            _positionInSuperStream = newPositionInSuperStream;
```

Comment:
Should we be calling _superStream.Seek here?  That has side-effects that are observable so I would have expected that a call to Seek in the SubStream would change the position in the _superStream immediately and not be deferred to some later read.